### ISSUE TITLE
Highly componentized design of environments

### ISSUE THREAD

**Issue by rusu24edward** [LLNL/Abmarl#12]
PredatorPreyEnv currently processes some interactions between the predators and prey. Each agent can move around on the grid and the predators can attack the prey. With #53, the prey can also harvest resources. This could be split up into 3 components: a movement handler, an agent-agent-interaction handler, and an agent-environment-interaction handler. This is a significant redesign, so it should be considered very carefully and not as part of the main development push.

The core of the environment is a grid with agents in cells. So each agent has a position.
Features are then layered onto this grid in the form of wrappers:
1. Movement (agent-env): Agents can choose to move. Env processes the move action. Agents have a movement parameter.
1. Resource harvest (agent-env): Resources is an addition to the environment because now there are values that exist on the grid. This is an update to the state, so it naturally works as a wrapper. Agents can choose to harvest. Env process the harvest action. Agents have a harvest parameter.
1. Attack (agent-agent): Agents can choose to attack each other. Env processes these actions. Agents have attack parameter.
1. Communicate (agent-agent): Agents can choose to share observations. This modifies the state of the env because now there is communication buffer, so this naturally works as a wrapper. Env processes action. Agents have communication parameters.

Two important questions:
1. How does the environment process the actions? Do they happen in order form outside in--communicate, then attack, then harvest, then move? Each action updating the state of the wrapper before going down to the next level?
1. What if we want to restrict actions from happening in the same step? Okay, we can define the action policy to only choose certain actions in this mutually exclusive way. But how do we define the exclusivity in the conglomeration of wrappers?

In Epic #10 

# Update
Using wrappers for each feature is poor design. We are using component-based design instead.

**Comment by rusu24edward** [comment]
Progress on this is coming along. We took the hackathon idea and extended it to a bunch of modules that work with different types of agents. These modules give the agents action and observation spaces and support different actions and returning different observations. Below is record of things that still need to be done before we can open a PR:

### Big Changes

Adding agent attributes via a hierarchy of classes is nice because it allows us
to work with agents as objects. However, we have to use multiple inheritance all
over the place, which results in some design contraints for the sake of MRO. MRO
is needed for classes because of all that they support, but we are not using classes
to their full support, so I don't need the same restrictions on the MRO. I just
need everything to be hit once. I searched around for this, and it looks like
dataclasses is a step in the right direction, but not quite because it doesn't
modify the MRO. Here are some links for doing so:
https://stackoverflow.com/questions/20822850/change-python-mro-at-runtime
http://pybites.blogspot.com/2009/01/mro-magic.html
Otherwise, if we don't want to use classes, we can probably just use dictionaries.
Every agent is a dictionary and they are configured if none of the values are None.
Using dictionary is a little annoying because we have to use the [''] instead of
just ., but maybe after the agent type is fully created I can use Box to convert
the dictionary to an object.

In movement, consider process_move to take the agent as input argument and update
the position of the agent. Restrictions can be put into the kwargs, such as squares
that are already occupied by other agents/obstacles.
The bigger idea here is to actually have the modules support an `act` function
that takes the acting agent as input and some kwargs for specifics. This streamlines
the modules that are affected by agent actions. Some modules, like resources, have
additional dynamics that are not affected by agent actions, such as the resources
regrowing; and DyingEnv doesn't update at all based on agent actions but only indirectly.
These dynamics can be captured in separate functions.
tl;dr: Streamline the Modules to implement an `act` function.


### Simple fixes

In attacking, include combination matrix in kwargs of process_attack that indicates
which agents can attack each other. This is similar to what we do in get_obs in
the communication wrapper.

In death_life, the process_death function also processes the max health bound,
which doesn't make sense. So we should create a function to update the agent's
health that caps the upper and lower bound.

After the agents action and observation spaces have been defined by each module,
we can wrap them with a gym Dict to ensure that the space is still a gym space.

There is probably a bug in the DyingAgent/Env because prey were being attacked 
a lot and were not dying. They may have been harvesting to replenish their health,
but we'll have to look into this more.

Assert that agents are configured at the end of init in the environment.


### Todo when we're done
Confirm that the environments are ABSi
Unit tests!

**Comment by rusu24edward** [comment]
- [ ] Change name of world to position since that's all it does

**Comment by rusu24edward** [comment]
Communication:
Communication is a subset of observations, a way to augment your observations with that of another agent. Typically we think of communication dynamics being point-to-point or distance-based. We can try to implement both of these within the grid observation framework.

Distance based: the agent will have a communication range that *should* be greater than its view. The agent sees the environment according to its view. Beyond that, the agent sees a "fog" until the end of its communication distance. Observations from other agents can fill in the fog with real data. We will need to modify the other agents observations in two ways, however. First, the obs from the sending agent must be translated to the correct location for the receiving agent's observation. Second, the obs from the sending agent may need to be clipped to fit within the receiving agent's total box.
For example, suppose `m` and `a` have a view of 2, and `m` has a total distance of 5. Suppose `a` is near `m` and an observation is exchanged. The following ascii art shows us that `a`'s observation, indicated with 1s, will be offset and partially clipped because the last column is outside of `a`'s total distance.
```
+ - - - - - - - - - - - +
| f f f f f f f f f f f |
| f f f f f f f f f f f |
| f f f f f f f f f f f |
| f f f 0 0 0 0 0 f f f |
| f f f 0 0 0 0 0 f f f |
| f f f 0 0 m 0 0 f f f + - +
| f f f 0 0 0 0 2 1 1 1 | 1 |
| f f f 0 0 0 0 2 1 1 1 | 1 |
| f f f f f f f 1 1 a 1 | 1 |
| f f f f f f f 1 1 1 1 | 1 |
| f f f f f f f 1 1 1 1 | 1 |
+ - - - - - - - - - - - + - +
```

**Comment by rusu24edward** [comment]
Following up on the above grid obs + communication comment, an easier path would be to use a fixed spatial observation, where all the agents observe the entire map. Most of the map is "fog" except for the patch near the agent. Then, when an observation is shared, a patch of the map is "unfogged" for that step.

**Comment by rusu24edward** [comment]
- [ ] Consider the design of the movement component

Design of the movement component is a bit ambiguous. On the one hand, it seems like the movement component should take the position and movement and output the desired new location, without constraints. Then, the environment can decide what to do about the movement output (e.g. if the new location is out of bounds, then don't move there). Currently, movement component does this constraint internally, checking for out of bounds movement, which it can do because it has the region. But as we add more constraints, such as landmarks and other agents' locations, the movement component will need to be updated to handle all these constraints. The question is: does it make sense for movement components to handle this internally, or is it better for movement to just process the desired location and let the environment or another component process the constraints? This has parallel considerations with the attack component with the question of processing the agents' health.

**Comment by rusu24edward** [comment]
- [ ] Consider a boundaries channel in observations

Position observations produce the relative position of other agents in the grid, and they also produce the relative position of boundaries. Perhaps it would be better to break down the boundaries observation as its own channel, separate from the agents (and from the resources).
